<div style="border:solid black 2px; padding: 20px">

# Reviewer's feedback v.1

**Halo Nasywa**
    
**Greetings from Chamdani**

Terima kasih telah mengirimkan project kamu dengan ini kamu sudah memulai sebuah langkah yang luar biasa saat ini. Disini saya akan mereview proyek kamu ya.

Saya akan memberikan beberapa komentar dan feedback seperti dibawah ini, **mohon jangan dipindah, dirubah, maupun dihapus ya :).**
    
> Mohon diperhatikan bahwa apabila ada temuan atau kesalahan yang sama namun tidak ada komentar perbaikan disana, itu berarti kamu masih tetap harus memperbaikinya yaa.

Komentar yang saya berikan akan muncul dalam warna hijau, kuning, atau pun merah seperti ini:

</div>

<div class="alert alert-success">
<b>Chamdani's comment v.*</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

<div class="alert alert-warning">
<b>Chamdani's comment v.*</b> <a class="tocSkip"></a>

Ada beberapa catatan.

</div>

<div class="alert alert-danger">
<b>Chamdani's comment v.*</b> <a class="tocSkip"></a>

Perlu beberapa perbaikan.

</div>

<div class="alert alert-info">
<b>Chamdani's info v.1</b> <a class="tocSkip"></a>

Catatan umum:
* Fungsi `print` tidak diperlukan untuk mencetak suatu nilai
* Lebih baik tulis kode pada bagian akhir cell code untuk menampilkan output tanpa `print`
* Perintah soal yang ada dalam tanda kurung siku `[...]` sebaiknya dihilangkan
* Sangat disarankan untuk menggunakan cell markdown untuk memberikan deskripsi temuan atau kesimpulan.
* Tidak disarankan menggunakan cell code untuk memberikan penjelasan / deskripsi.
* Untuk menghindari galat pada notebook project kamu. Batasi jumlah output yang dikeluarkan setiap kode dengan batas max 10 baris data.

</div>

<div style="border:solid gray 5px; padding: 20px" class="alert alert-warning">

**INGATLAH** 
* Project kamu tidak akan bisa diterima bila kamu masih memiliki feedback berwarna merah. 
* Kamu hanya memiliki 3 kali kesempatan iterasi submission disetiap projectnya 
* Kamu dapat memperbaiki dan menyelesaikan feedback **pertama** dari saya tidak lebih dari 1\*24 jam setiap iterasinya. *Untuk membantu saya mengetahui apakah ada kesulitan dari feedback project kamu*
* Jika merasa kesulitan kamu dapat menuliskan responmu atas feedback saya menggunakan blok markdown warna biru yaa.

Kamu dapat menjawab saya dengan menggunakan blok berwarna biru seperti ini

<div class="alert alert-info">
<b>Student answer</b> <a class="tocSkip"></a>

Siap kak.

</div>

<div style="border:solid black 2px; padding: 20px" class="alert alert-success">
    
# General feedback v.1 *Project Accepted*

Terima kasih banyak telah mengirimkan proyek Anda! 
<br>
    
Secara keseluruhan, proyek Anda **SANGAT** mengesankan. Semoga apa yang kamu pelajari dalam project ini dapat membantu meningkatkan kemampuanmu. Pertahankan apa yang sudah bagus, dan tingkatkan apa yang menurutmu masih perlu ditingkatkan. Semoga berhasil pada project selanjutnya yaa :)

**TETAP SEMANGAT!!**
    
</div>

## Deskripsi Proyek

Saya adalah analis OilyGiant yang ingin memprediksi lokasi yang cocok untuk penggalian sumur minyak yang baru. Saya akan menggunakan data sampel minyak dari tiga wilayah. Model yang dibuat akan membantu OilyGiant untuk memilih wilayah dengan margin laba tertinggi.

## Tujuan Proyek

Tujuan proyek ini adalah menyarankan suatu wilayah untuk pengembangan sumur minyak dengan menyertakan pembenaran atau alasan pemilihan wilayah. 
Dalam proses ini, akan dilakukan bootstrapping dengan 1.000 sampel untuk menemukan distribusi labanya. Dan juga akan dilakukan penghitungan laba dan risiko untuk setiap wilayah.

## Eksplorasi Dataset

Pada tahap ini, akan dilakukan eksplorasi dataset dan juga mendeskripsikan kolom-kolom pada dataset.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Memuat Library

In [1]:
# pengolahan dataset
import pandas as pd, numpy as np
# library untuk scientific computing
import numpy as np

from numpy.random import RandomState

# import modul LinearRegression dari library scikit learn
from sklearn.linear_model import LinearRegression

# import modul Train Test Split untuk membagi data dari library scikit learn
from sklearn.model_selection import train_test_split

# import module shuffle agar proses shuffle array dilakukan secara konsisten
from sklearn.utils import shuffle

# import module mean squared error untuk penghitungan rata-rata Kesalahan kuadrat antara nilai aktual dan nilai peramalan.
from sklearn.metrics import mean_squared_error

# Library untuk mengetahui progress ketika proses bootstrapping
from tqdm import tqdm

# ignore warning
import warnings
from pandas.core.common import SettingWithCopyWarning
warnings.filterwarnings("ignore")

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Memuat Dataset

In [2]:
# path dataset
path_df_wilayah0 = '/datasets/geo_data_0.csv'
path_df_wilayah1 = '/datasets/geo_data_1.csv'
path_df_wilayah2 = '/datasets/geo_data_2.csv'

In [3]:
# load dataset
df0 = pd.read_csv(path_df_wilayah0)
df1 = pd.read_csv(path_df_wilayah1)
df2 = pd.read_csv(path_df_wilayah2)

In [4]:
df0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [5]:
df1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [6]:
df2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Kolom

**Fitur**
- `id` - ID unik sumur minyak
- `f0, f1, f2` - tiga fitur titik

**Target**
- `product` - volume cadangan minyak di sumur (ribuan barel).

### Deskripsi Data

#### df0

In [7]:
# Mengetahui jumlah rows dan kolom pada dataset df0
df0.shape

(100000, 5)

In [8]:
# Mengetahui informasi jumlah rows pada masing-masing kolom dan tipe datanya
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [9]:
# Pengecekan apakah ada missing values pada masing-masing kolom
df0.isnull().sum() / 100

id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64

In [10]:
# Mengetahui informasi statistik dari kolom numerical pada dataset
df0.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

**Kesimpulan tahap Eksplorasi Dataset df0**

1. Tidak ada missing values pada df0
2. Tipe data pada kolom sudah benar
3. Akan dilakukan drop pada kolom id karena tidak digunakan.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

#### df1

In [11]:
# Mengetahui jumlah rows dan kolom pada dataset df1
df1.shape

(100000, 5)

In [12]:
# Mengetahui informasi jumlah rows pada masing-masing kolom dan tipe datanya
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [13]:
# Pengecekan apakah ada missing values pada masing-masing kolom
df1.isnull().sum() / 100

id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64

In [14]:
# Mengetahui informasi statistik dari kolom numerical pada dataset
df1.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

**Kesimpulan tahap Eksplorasi Dataset df1**

1. Tidak ada missing values pada df1
2. Tipe data pada kolom sudah benar
3. Akan dilakukan drop pada kolom id karena tidak digunakan.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

#### df2

In [15]:
# Mengetahui jumlah rows dan kolom pada dataset df2
df2.shape

(100000, 5)

In [16]:
# Mengetahui informasi jumlah rows pada masing-masing kolom dan tipe datanya
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [17]:
# Pengecekan apakah ada missing values pada masing-masing kolom
df2.isnull().sum() / 100

id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64

In [18]:
# Mengetahui informasi statistik dari kolom numerical pada dataset
df2.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

**Kesimpulan tahap Eksplorasi Dataset df2**

1. Tidak ada missing values pada df2
2. Tipe data pada kolom sudah benar
3. Akan dilakukan drop pada kolom id karena tidak digunakan.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Drop kolom yang tidak digunakan dan penggabungan ketiga wilayah ke 1 dataset

In [19]:
all_data = [
    df0.drop('id', axis =1),
    df1.drop('id', axis =1),
    df2.drop('id', axis =1),
]

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Memisahkan Data dan Melatih Model

In [20]:
state = RandomState(42)

samples_target = []
samples_prediction = []

for wilayah in range(len(all_data)):
    data = all_data[wilayah]
    
    features = data.drop('product', axis = 1)
    target = data['product']
    # split training dan testing
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state = state)
    
    # training model
    # definisi model Linear Regression
    model = LinearRegression()
    # latih model pada training set
    model.fit(features_train, target_train)
    # membuat prediksi menggunakan training set
    predictions = model.predict(features_valid)
    
    samples_target.append(target_valid.reset_index(drop=True))
    samples_prediction.append(pd.Series(predictions))

    # Mengetahui root mean squared error pada training dataset
    rmse_train = np.sqrt(mean_squared_error(target_valid, predictions))
    # Average Barrel
    avg_product_target = target.mean()
    avg_product_predictions = predictions.mean()
    
    # menampilkan hasil
    print('--------------------------------')
    print(f'Training Result Geo Data {wilayah}')
    print(f'Average Barrel Target: {avg_product_target}')
    print(f'Average Barrel Predictions: {avg_product_predictions}')
    print(f'RMSE Training: {rmse_train}')

--------------------------------
Training Result Geo Data 0
Average Barrel Target: 92.50000000000001
Average Barrel Predictions: 92.3987999065777
RMSE Training: 37.756600350261685
--------------------------------
Training Result Geo Data 1
Average Barrel Target: 68.82500000000002
Average Barrel Predictions: 68.58780312701154
RMSE Training: 0.8899472027156135
--------------------------------
Training Result Geo Data 2
Average Barrel Target: 95.00000000000004
Average Barrel Predictions: 95.09230341410031
RMSE Training: 40.23735038140562


**Hasil Analisis:**

1. Jika dilihat dari rata-rata cadangan product atau minyak, maka Wilayah 2 memiliki nilai paling tinggi.
2. Jika dilihat dari nilai RMSE, dapat dilihat bahwa df1 memiliki RMSE paling kecil (range average error). Hal ini menggambarkan bahwa angka pasti nilai average barrel yaitu  diantara 67.81 (68.72-0.89) sampai dengan 69.7 (68.72+0.89)

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Persiapan Penghitungan Laba

**Tahapan Persiapan**

1. Penyimpanan nilai kunci untuk perhitungan laba pada variabel-variabel terpisah.

In [22]:
# penyimpanan nilai kunci
product_price = 4500
total_cost = 100000000 # 100 juta USD
points = 200
cost_per_point = total_cost/points
point_per_cost = int(total_cost / cost_per_point)
point_per_cost

200

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Menghitung Profit Pada 200 titik tertinggi

**Proses yang akan dilakukan pada tahap ini adalah sebagai berikut:**

Pembuatan fungsi untuk mengetahui:
- profit paling tinggi dari ketiga wilayah
- total cost, total income
- total cadangan minyak
- sumur dengan nilai prediksi paling tinggi

Dalam pembuatan fungsi, akan mempertimbangkan hal-hal berikut ini:

a. Hanya akan dipilih top 200 sumur

b. Anggaran untuk mengembangkan 200 sumur minyak adalah 100 juta USD. (total_cost)

c. Income adalah 4500 per 1000 barel (income)

Kemudian, akan dihitung target (real) volume cadangan minyak berdasarkan prediksi-prediksi yang telah dibuat.

### Pembuatan Fungsi untuk menghitung profit pada data prediksi

In [24]:
# fungsi untuk menghitung laba dari kumpulan sumur terpilih pada data prediksi
def profitability(prediction, name, income=4500, total_cost = 100000000, points = 200):
    prediction = prediction[name]
    prediction_top200 = prediction.sort_values(ascending=False)[:points].reset_index()
    total_minyak = prediction_top200[0].sum()
    total_cost = round(total_cost/1000000)
    total_income = round(income * total_minyak/1000000)
    profit = round(total_income - total_cost)
    
    # menampilkan hasil
    print('-------------------------------')
    print(f'Profitability Geo Data: {name}')
    print(f'Profit: {profit}', 'M USD')
    print(f'Total Cost: {total_cost}', 'M USD')
    print(f'Total Income: {total_income}', 'M USD')
    print(f'Total Cadangan Minyak: {total_minyak}')

In [25]:
# Implementasi fungsi untuk mengetahui profit dan cadangan minyak
profitability(prediction=samples_prediction, name = 0)
profitability(prediction=samples_prediction, name = 1)
profitability(prediction=samples_prediction, name = 2)

-------------------------------
Profitability Geo Data: 0
Profit: 39 M USD
Total Cost: 100 M USD
Total Income: 139 M USD
Total Cadangan Minyak: 30881.463288146995
-------------------------------
Profitability Geo Data: 1
Profit: 25 M USD
Total Cost: 100 M USD
Total Income: 125 M USD
Total Cadangan Minyak: 27746.394260320503
-------------------------------
Profitability Geo Data: 2
Profit: 35 M USD
Total Cost: 100 M USD
Total Income: 135 M USD
Total Cadangan Minyak: 29941.147885588958


**Kesimpulan**

1. Dari hasil fungsi yang telah dibuat, dapat diketahui bahwa Wilayah Geo Data 0 memiliki profit paling tinggi, total income paling tinggi, total cadangan minyak paling tinggi, dan volume cadangan minyak pada sumur (yang paling tinggi) tertinggi juga.
2. Hasil ini belum final, dikarenakan harus mempertimbangkan nilai RMSE pada proses sebelumnya, dimana wilayah Geo Data 0 memiliki risiko yang tinggi 37%.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Pembuatan Fungsi untuk menghitung profit pada data real berdasrkan hasil prediksi top200

In [26]:
# mengambil prediksi top 200 sumur minyak
top200_predictions = samples_prediction[0].sort_values(ascending=False)[:point_per_cost]

In [27]:
top200_predictions

6958     176.536104
18194    176.274510
17251    173.249504
457      172.802708
2202     172.744977
            ...    
8772     146.952653
15975    146.948764
7846     146.947483
6633     146.946515
7154     146.874947
Length: 200, dtype: float64

In [28]:
# mencocokkan index pada hasil prediksi dengan data samples_target
top200_target = samples_target[0][top200_predictions.index]
top200_target

6958     153.639837
18194    140.631646
17251    178.879516
457      176.807828
2202     130.985681
            ...    
8772     132.951877
15975    136.027691
7846     162.142530
6633     120.536962
7154     138.424174
Name: product, Length: 200, dtype: float64

In [29]:
# penghitungan jumlah total product atau minyak
total_product = top200_target.sum()
total_product

29686.9802543604

In [30]:
# penghitungan revenue
revenue = total_product*4500
revenue

133591411.14462179

In [31]:
# penghitungan profit
profit = revenue - total_cost
profit

33591411.14462179

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

**Dari langkah-langkah penghitungan profit real diatas, dibuat fungsi untuk menghitung profit, seperti dibawah:**

In [32]:
# fungsi untuk menghitung profit

def target_profit(target, predictions):
    predictions_sorted = predictions.sort_values(ascending=False)
    selected_points = target[predictions_sorted.index][:point_per_cost]
    product = selected_points.sum()
    revenue = product * product_price
    profit = revenue - total_cost
    return profit

In [33]:
# menerapkan fungsi pada masing-masing wilayah
print('Profit real berdasarkan hasil prediksi pada wilayah 0:',target_profit(samples_target[0], samples_prediction[0]))
print('Profit real berdasarkan hasil prediksi pada wilayah 1:',target_profit(samples_target[1], samples_prediction[1]))
print('Profit real berdasarkan hasil prediksi pada wilayah 2:',target_profit(samples_target[2], samples_prediction[2]))

Profit real berdasarkan hasil prediksi pada wilayah 0: 33591411.14462179
Profit real berdasarkan hasil prediksi pada wilayah 1: 24150866.966815114
Profit real berdasarkan hasil prediksi pada wilayah 2: 25841568.87152408


**Kesimpulan**

1. Dari hasil fungsi yang telah dibuat, dapat diketahui bahwa Wilayah Geo Data 0 pada data real memiliki profit paling tinggi, yaitu sebanyak 34 M USD.
2. Hasil ini belum final, dikarenakan harus mempertimbangkan nilai RMSE pada proses sebelumnya, dimana wilayah Geo Data 0 memiliki risiko yang tinggi 37%.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Perbandingan Nilai Profit pada data prediksi dan data real

Pada data prediksi, nilai profit pada wilayah 0 adalah 40M USD. Sedangkan, jika menggunakan data real(samples_target) nilai profit pada wilayah 0 adalah 33M USD.
Sehingga, ada kesalahan sekitar 7M pada data prediksi.

## Penghitungan Risiko dan Laba dengan menggunakan Teknik Bootstrapping

**Kondisi**

1. Dalam penggunaan teknik bootstrapping, akan diambil samplek top 500, dengan dilakukan pengulangan sebanyak 1,000 kali. 
2. Dari hasil bootstrapping, maka akan diambil top 200 (200 titik terbaik) untuk penghitungan laba.
3. Selang kepercayaan yang digunakan adalah 95%, dengan mengambil 2,5% dari ujuang atas dan ujung bawah
4. Pertahankan hanya wilayah dengan risiko kerugian lebih rendah dari 2,5%. Dari daftar wilayah yang memenuhi kriteria, pilihlah wilayah dengan rata-rata laba tertinggi.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Pembuatan Fungsi untuk menghitung profit

In [34]:
sample_size = 500
bootstrap_size = 1000

In [35]:
def calculate_profit_bootstrap(prediction, name, income=4500, total_cost = 100000000, points = 200):
    predict_top200 = prediction.sort_values(ascending=False)[:points]
    product = predict_top200.sum()
    total_income = income * product
    profit = total_income - total_cost
    return profit

In [36]:
# fungsi untuk menghitung profit

def profit(target, predictions):
    predictions_sorted = predictions.sort_values(ascending=False)
    selected_points = target[predictions_sorted.index][:point_per_cost]
    product = selected_points.sum()
    revenue = product * product_price
    profit = revenue - total_cost
    return profit

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Penghitungan profit pada data prediksi menggunakan teknik bootstrapping

In [37]:
for wilayah in range(3):
    
    target = samples_target[wilayah]
    predictions = samples_prediction[wilayah]
    
    profit_values = []
    
    for i in tqdm(range(bootstrap_size)):
        target_sample= target.sample(n=sample_size, replace = True, random_state = state)
        prediction_sample= predictions[target_sample.index]
        profit_values.append(calculate_profit_bootstrap(prediction= prediction_sample, name=wilayah))
        
    profit_values = pd.Series(profit_values)
    
    mean_profit = profit_values.mean()
    confidence_interval = (profit_values.quantile(0.025), profit_values.quantile(0.975))
    negative_profit_chance = (profit_values <0).mean()
    
    print('-- Wilayah', wilayah, '--')
    print('Mean Profit: ', mean_profit, 'USD')
    print('95% confidence interval:', confidence_interval)
    print('Risk of losses :', negative_profit_chance)
    print()

100%|██████████| 1000/1000 [00:00<00:00, 1211.01it/s]


-- Wilayah 0 --
Mean Profit:  3324499.1511043045 USD
95% confidence interval: (1010833.3330253005, 5593445.146503954)
Risk of losses : 0.002



100%|██████████| 1000/1000 [00:00<00:00, 1110.44it/s]


-- Wilayah 1 --
Mean Profit:  4269058.93421017 USD
95% confidence interval: (348183.1922796317, 8115279.249680298)
Risk of losses : 0.016



100%|██████████| 1000/1000 [00:00<00:00, 1171.94it/s]

-- Wilayah 2 --
Mean Profit:  3017845.5962736136 USD
95% confidence interval: (1167160.831588246, 5002914.4776899135)
Risk of losses : 0.001



**Hasil Analisis**

Jika menggunakan data prediksi untuk menghitung Profit, maka rata-rata profit tertinggi ada pada wilayah 1, meskipun risk of loses juga paling tinggi diantara ketiga wilayah.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Penghitungan profit pada data real menggunakan teknik bootstrapping

In [38]:
for wilayah in range(3):
    
    target = samples_target[wilayah]
    predictions = samples_prediction[wilayah]
    
    profit_values = []
    
    for i in tqdm(range(bootstrap_size)):
        target_sample= target.sample(n=sample_size, replace = True, random_state = state)
        prediction_sample= predictions[target_sample.index]
        profit_values.append(target_profit(target_sample, prediction_sample))
        
    profit_values = pd.Series(profit_values)
    
    mean_profit = profit_values.mean()
    confidence_interval = (profit_values.quantile(0.025), profit_values.quantile(0.975))
    negative_profit_chance = (profit_values <0).mean()
    
    print('-- Wilayah', wilayah, '--')
    print('Mean Profit: ', mean_profit, 'USD')
    print('95% confidence interval:', confidence_interval)
    print('Risk of losses :', negative_profit_chance)
    print()

100%|██████████| 1000/1000 [00:01<00:00, 594.26it/s]


-- Wilayah 0 --
Mean Profit:  4341456.987107678 USD
95% confidence interval: (-576180.0642516082, 9645925.33771113)
Risk of losses : 0.049



100%|██████████| 1000/1000 [00:01<00:00, 604.86it/s]


-- Wilayah 1 --
Mean Profit:  4841242.252647654 USD
95% confidence interval: (488526.54075440543, 8966394.898976061)
Risk of losses : 0.014



100%|██████████| 1000/1000 [00:01<00:00, 598.06it/s]

-- Wilayah 2 --
Mean Profit:  3451884.409861117 USD
95% confidence interval: (-1898801.4969076396, 8989624.74503065)
Risk of losses : 0.116



**Hasil Analisis**

Jika menggunakan data real/target untuk menghitung Profit, maka rata-rata profit tertinggi ada pada wilayah 1 dengan risk of loses paling kecil, yaitu 0.014.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Perbandingan Nilai Profit pada data prediksi dan data real dengan menggunakan teknik bootstrapping

Nilai rata-rata profit tertinggi ada di wilayah 0, baik menggunakan data prediksi maupun data real tertinggi, meskipun tingkat risk of loses berbeda.

## Kesimpulan

Hasil dari masing-masing proses:

1. Pada tahap awal, dilakukan penghitungan tingkat akurasi hasil perkiraan model Linear Regression, wilayah 1 (37.75) memiliki nilai RMSE terendah (0.89), dan memiliki perbedaan yang cukup jauh dengan wilayah 0 dan wilayah 2 (40.23). Hal ini menggambarkan bahwa angka pasti nilai average barrel yaitu diantara 67.81 (68.72-0.89) sampai dengan 69.7 (68.72+0.89)
2. Penghitungan profit tanpa teknik bootstrapping, jika menggunakan data prediksi, wilayah 0 memiliki nilai tertinggi dengan nilai profit adalah 40M. Sedangkan, jika menggunakan data real(samples_target) nilai profit pada wilayah 0 adalah 34M. Sehingga, ada kesalahan sekitar 6M pada data prediksi.
3. Penghitungan profit tertinggi menggunakan teknik boostrapping, jika menggunakan data prediksi, wilayah 1 memiliki nilai rata-rata profit tertinggi. Nilai rata-rata profit tertinggi jika menggunakan data real juga ada di wilayah 1 dengan nilai risk of loses (nilai profit yang mengalami kerugian atau minus) paling kecil . 

**Wilayah terbaik berdasarkan hasil analiiss**

Dari hasil analisis, **Wilayah 1** merupakan wilayah terbaik untuk mengembangkan sumur terbaru. 
Hal ini didasarkan nilai RMSE yang kecil 0.89 dan juga risiko kerugian 0.016 yang dibawah tingkat yang diterapkan (2.5%)

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>